commons/src/lib.rs

```diff
@@ -1,15 +1,93 @@
+#[macro_use]
+extern crate failure;
+extern crate url;
+
+use failure::Fallible;
+use std::collections::HashSet;
+use url::form_urlencoded;
+
 /// Strip all but one leading slash and all trailing slashes
 pub fn parse_path_prefix(path_prefix: &str) -> String {
     format!("/{}", path_prefix.to_string().trim_matches('/'))
 }
 
+/// Parse a comma-separated set of client parameters keys.
+pub fn parse_params_set(params: &str) -> HashSet<String> {
+    params
+        .split(',')
+        .into_iter()
+        .filter_map(|key| {
+            let trimmed = key.trim().to_string();
+            if trimmed.is_empty() {
+                None
+            } else {
+                Some(trimmed)
+            }
+        })
+        .collect()
+}
+
+/// Make sure `query` string contains all `params` keys.
+pub fn ensure_query_params(params: &HashSet<String>, query: &str) -> Fallible<()> {
+    // No mandatory parameters, always fine.
+    if params.is_empty() {
+        return Ok(());
+    }
+
+    // Extract and de-duplicate keys from input query.
+    let query_keys: HashSet<String> = form_urlencoded::parse(query.as_bytes())
+        .into_owned()
+        .map(|(k, _)| k)
+        .collect();
+
+    // Make sure all mandatory parameters are present.
+    ensure!(
+        params.is_subset(&query_keys),
+        "mandatory parameters missing"
+    );
+
+    Ok(())
+}
+
 #[cfg(test)]
 mod tests {
+    use super::*;
+
     #[test]
     fn test_parse_path_prefix() {
-        assert_eq!(super::parse_path_prefix("//a/b/c//"), "/a/b/c");
```

Comment:
Would you mind separating this refactor into a preceding commit? Just in the case of reverting the new feature we can keep the other.